task/sensor-api/build.rs

```diff
@@ -2,7 +2,132 @@
 // License, v. 2.0. If a copy of the MPL was not distributed with this
 // file, You can obtain one at https://mozilla.org/MPL/2.0/.
 
-fn main() -> Result<(), Box<dyn std::error::Error>> {
-    idol::client::build_client_stub("../../idl/sensor.idol", "client_stub.rs")?;
+use anyhow::{anyhow, bail, Result};
+use serde::Deserialize;
+use std::collections::{BTreeMap, BTreeSet};
+use std::{fmt::Write as FmtWrite, io::Write as IoWrite};
+
+/// This represents our _subset_ of global config and _must not_ be marked with
+/// `deny_unknown_fields`!
+#[derive(Deserialize)]
+#[serde(rename_all = "kebab-case")]
+struct GlobalConfig {
+    sensor: Option<SensorConfig>,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "kebab-case", deny_unknown_fields)]
+struct SensorConfig {
+    devices: Vec<Sensor>,
+}
+
+#[derive(Debug, Deserialize)]
+#[serde(rename_all = "kebab-case", deny_unknown_fields)]
+struct Sensor {
+    name: String,
+    device: String,
+    description: String,
+    sensors: BTreeMap<String, usize>,
+}
+
+fn main() -> Result<()> {
+    idol::client::build_client_stub("../../idl/sensor.idol", "client_stub.rs")
+        .map_err(|e| anyhow!("idol error: {e}"))?;
+
+    build_i2c::codegen(build_i2c::Disposition::Sensors)?;
+
+    let config: GlobalConfig = build_util::config()?;
+
+    let (count, text) = if let Some(config_sensor) = &config.sensor {
+        let sensor_count: usize =
+            config_sensor.devices.iter().map(|d| d.sensors.len()).sum();
+
+        let mut by_device: BTreeMap<String, Vec<usize>> = BTreeMap::new();
+        let mut names = BTreeSet::new();
+        for (i, d) in config_sensor.devices.iter().enumerate() {
+            by_device.entry(d.device.clone()).or_default().push(i);
+            if !names.insert(d.name.clone()) {
+                bail!("Duplicate sensor name: {}", d.name);
+            }
+        }
+
+        let mut sensors_text = String::new();
+        let mut sensor_id = 0;
+        for d in &config_sensor.devices {
```

Comment:
I'm not sure this is the right place for this code - in particular, I assume we (eventually?) want to be able to report these sensor readings up to MGS, which means `control-plane-agent` needs to be able to step over this list at runtime. For i2c sensors I mucked with `build-i2c` to add sensor descriptions to `validate-api`, but it seems likely that's not really the right place either? 

I don't have a concrete suggestion here and leave this comment primarily as a data point going into the upcoming discussion on organizing sensors.